### DISCUSSION TITLE
7. UDP의 오류 검출 방법에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by eun-byeol** [jewelry-box/cs-interview#12] [General]
UDP의 오류 검출 방법에 대해 설명해 주세요.
체크섬으로 데이터 유실을 어떻게 판단하나요?

**[키워드]**
- UDP의 오류 검출
- UDP 체크섬

**Comment by cutehumanS2** [comment]
UDP는 헤더의 **체크섬 필드**를 사용하여 데이터 전송 중 발생할 수 있는 오류를 검출합니다. 
체크섬은 **데이터 무결성을 확인하는 최소한의 검증 수단**으로, 송신 측에서 계산된 체크섬 값과 수신 측에서 계산된 체크섬 값의 비교를 통해 데이터 손상 여부를 판단할 수 있습니다.
다만, 패킷 자체의 유실 여부는 감지할 수 없으며, 손실된 데이터는 복구할 수 없다는 한계가 있어, 
UDP는 신뢰성이 요구되는 데이터 전송보다는, 속도가 중요한 실시간 스트리밍 등에서 더 적합하게 사용됩니다.

**Comment by cutehumanS2** [comment]
### UDP 체크섬
UDP 체크섬은 데이터 전송 중 발생할 수 있는 오류를 검출하기 위해 사용되는 메커니즘입니다. UDP는 비신뢰성 프로토콜이지만, 최소한의 데이터 무결성을 보장하기 위해 체크섬을 사용합니다.

UDP 체크섬은 전송된 데이터그램이 손상되지 않았는지 확인하는 데 사용되며, 송신 측에서 계산된 체크섬 값과 수신 측에서 계산된 체크섬 값이 일치하지 않을 경우 해당 데이터그램이 전송 중에 손상되었다고 간주하여 폐기합니다.

### UDP 체크섬 계산 과정
1. 의사 헤더(발신 IP 주소, 수신 IP 주소, 프로토콜 번호, UDP 길이 등), UDP 헤더, 데이터를 16비트로 쪼개어 모두 더합니다.
2. 오버플로나 캐리 아웃(carry-out)된 비트를 다시 더합니다.
3. 1의 보수를 취해 체크섬 값을 만듭니다.
<img width="545" alt="image" src="https://github.com/user-attachments/assets/57869647-cf39-4109-8515-0c523ec708a9" />
<img width="526" alt="image" src="https://github.com/user-attachments/assets/2ccf712c-fc50-438d-8d01-d839b30ddfcd" />


### 오류 검출 과정
1. **송신 측**
    - 송신 측은 UDP 헤더, 데이터, 의사 헤더(발신 IP 주소, 수신 IP 주소, 프로토콜 번호, UDP 길이 등)로 체크섬 값을 계산합니다.
    - 계산된 체크섬 값은 UDP 헤더의 체크섬 필드에 저장하여 함께 전송합니다.
2. **수신 측**
    - 수신 측은 UDP 헤더, 데이터, 의사 헤더를 사용해 체크섬 값을 다시 계산합니다.
    - 계산한 값과 UDP 헤더에 저장된 체크섬 값을 비교합니다.
        - 수신된 체크섬을 포함하여 모든 값을 더한 결과가 0xFFFF(1의 보수에서 모든 비트가 1)이면 데이터가 정상적으로 전달된 것으로 간주합니다.
        - 결과가 0xFFFF가 아닌 경우, 데이터가 손상된 것으로 판단하여 해당 패킷을 폐기합니다.

### 체크섬 방식의 한계점
- **오류를 100% 검출할 수 없습니다**: 데이터를 각각 비교하지 않고 값을 더해 검증하므로, 데이터의 순서가 바뀌거나 오류가 발생해도 체크섬이 동일하게 계산될 가능성이 있습니다.
- **패킷이 유실된 경우는 감지할 수 없습니다:** 체크섬은 데이터의 손상 여부만 확인할 수 있고, 네트워크에서 패킷 자체가 유실된 경우에는 이를 감지하지 못합니다.
- **손실된 데이터를 복구할 수 없습니다**: UDP는 오류가 발생하면 해당 패킷은 폐기되며, 복구를 위한 재전송 기능을 제공하지 않습니다.
- **선택적 필드이기 때문에 체크섬을 사용하지 않으면 오류를 검출할 방법이 없습니다:** UDP 헤더의 체크섬 필드가 0으로 설정되면 검증이 비활성화되며, 데이터 손상 여부를 확인할 수 없습니다.
    - +) IPv4에서는 체크섬을 사용하지 않는 것이 가능하지만, IPv6에서는 패킷의 무결성 확인을 위해 반드시 체크섬을 사용해야 합니다.

<br><br>

[참고한 자료]
- https://whitehartlane.tistory.com/60
- https://roughroad.tistory.com/543
- https://velog.io/@dpf226/UDP-%EC%95%8C%EC%95%84%EB%B3%B4%EA%B8%B0
- https://www.brainkart.com/article/User-Datagram-Protocol-%28UDP%29_13485/ - 이미지
